File: smallpict-java/src/main/java/com/smallpict/models/OptimizeOptions.java
package com.smallpict.models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

@JsonInclude(JsonInclude.Include.NON_NULL)
public class OptimizeOptions {
    private final ImageFormat format;
    private final Integer quality;
    @JsonProperty("max_width")
    private final Integer maxWidth;
    @JsonProperty("max_height")
    private final Integer maxHeight;
    private final FitMode fit;
    private final Boolean lossless;
    @JsonProperty("strip_metadata")
    private final boolean stripMetadata;
    private final String filename;
    @JsonProperty("mime_type")
    private final String mimeType;
    @JsonIgnore
    private final String idempotencyKey;

    private OptimizeOptions(Builder builder) {
        this.format = builder.format;
        this.quality = builder.quality;
        this.maxWidth = builder.maxWidth;
        this.maxHeight = builder.maxHeight;
        this.fit = builder.fit;
        this.lossless = builder.lossless;
        this.stripMetadata = builder.stripMetadata;
        this.filename = builder.filename;
        this.mimeType = builder.mimeType;
        this.idempotencyKey = builder.idempotencyKey;
    }

    public static Builder builder() {
        return new Builder();
    }

    public ImageFormat getFormat() {
        return format;
    }

    public Integer getQuality() {
        return quality;
    }

    public Integer getMaxWidth() {
        return maxWidth;
    }

    public Integer getMaxHeight() {
        return maxHeight;
    }

    public FitMode getFit() {
        return fit;
    }

    public Boolean getLossless() {
        return lossless;
    }

    public boolean isStripMetadata() {
        return stripMetadata;
    }

    public String getFilename() {
        return filename;
    }

    public String getMimeType() {
        return mimeType;
    }

    public String getIdempotencyKey() {
        return idempotencyKey;
    }

    public static class Builder {
        private ImageFormat format = ImageFormat.AUTO;
        private Integer quality = 80;
        private Integer maxWidth;
        private Integer maxHeight;
        private FitMode fit = FitMode.COVER;
        private Boolean lossless = false;
        private boolean stripMetadata = true;
        private String filename;
        private String mimeType;
        private String idempotencyKey;

        public Builder format(ImageFormat format) {
            this.format = format;
            return this;
        }

        public Builder quality(int quality) {
            this.quality = Math.max(1, Math.min(100, quality));
            return this;
        }

        public Builder maxWidth(int maxWidth) {
            this.maxWidth = maxWidth;
            return this;
        }

        public Builder maxHeight(int maxHeight) {
            this.maxHeight = maxHeight;
            return this;
        }

        public Builder fit(FitMode fit) {
            this.fit = fit;
            return this;
        }

        public Builder lossless(boolean lossless) {
            this.lossless = lossless;
            return this;
        }

        public Builder stripMetadata(boolean stripMetadata) {
            this.stripMetadata = stripMetadata;
            return this;
        }

        public Builder filename(String filename) {
            this.filename = filename;
            return this;
        }

        public Builder mimeType(String mimeType) {
            this.mimeType = mimeType;
            return this;
        }

        public Builder idempotencyKey(String idempotencyKey) {
            this.idempotencyKey = idempotencyKey;
            return this;
        }

        public OptimizeOptions build() {
            return new OptimizeOptions(this);
        }
    }
}

File: smallpict-php/src/Models/OptimizeOptions.php
<?php

declare(strict_types=1);

namespace SmallPict\Models;

class OptimizeOptions
{
    private string $format;
    private int $quality;
    private ?int $maxWidth;
    private ?int $maxHeight;
    private string $fit;
    private bool $lossless;
    private bool $stripMetadata;
    private ?string $filename;
    private ?string $mimeType;
    private ?string $idempotencyKey;

    public function __construct(
        string $format = ImageFormat::AUTO,
        int $quality = 80,
        ?int $maxWidth = null,
        ?int $maxHeight = null,
        string $fit = FitMode::COVER,
        bool $lossless = false,
        bool $stripMetadata = true,
        ?string $filename = null,
        ?string $mimeType = null,
        ?string $idempotencyKey = null
    ) {
        $this->format = $format;
        $this->quality = max(1, min(100, $quality));
        $this->maxWidth = $maxWidth;
        $this->maxHeight = $maxHeight;
        $this->fit = $fit;
        $this->lossless = $lossless;
        $this->stripMetadata = $stripMetadata;
        $this->filename = $filename;
        $this->mimeType = $mimeType;
        $this->idempotencyKey = $idempotencyKey;
    }

    public function getFormat(): string
    {
        return $this->format;
    }

    public function getQuality(): int
    {
        return $this->quality;
    }

    public function getMaxWidth(): ?int
    {
        return $this->maxWidth;
    }

    public function getMaxHeight(): ?int
    {
        return $this->maxHeight;
    }

    public function getFit(): string
    {
        return $this->fit;
    }

    public function isLossless(): bool
    {
        return $this->lossless;
    }

    public function shouldStripMetadata(): bool
    {
        return $this->stripMetadata;
    }

    public function getFilename(): ?string
    {
        return $this->filename;
    }

    public function getMimeType(): ?string
    {
        return $this->mimeType;
    }

    public function getIdempotencyKey(): ?string
    {
        return $this->idempotencyKey;
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'format' => $this->format,
            'quality' => $this->quality,
            'max_width' => $this->maxWidth,
            'max_height' => $this->maxHeight,
            'fit' => $this->fit,
            'lossless' => $this->lossless,
            'strip_metadata' => $this->stripMetadata,
        ];
    }
}
